File: practice/ICPC2016Divisional/I.py
# I
class suffix:
    def __init__(self):
        self.index = 0
        self.rank = [0, 0]


def buildSuffix(txt, n):
    s = [suffix() for _ in range(n)]
    for i in range(n):
        s[i].index = i
        s[i].rank[0] = (ord(txt[i]) - ord("A"))
        s[i].rank[1] = (ord(txt[i + 1]) - ord("A")) if ((i + 1) < n) else -1
    s = sorted(s, key=lambda x: (x.rank[0], x.rank[1]))
    ind = [0] * n
    k = 4
    while k < 2 * n:
        rank = 0
        prev_rank = s[0].rank[0]
        s[0].rank[0] = rank
        ind[s[0].index] = 0
        for i in range(1, n):
            if s[i].rank[0] == prev_rank and s[i].rank[1] == s[i - 1].rank[1]:
                prev_rank = s[i].rank[0]
                s[i].rank[0] = rank
            else:
                prev_rank = s[i].rank[0]
                rank += 1
                s[i].rank[0] = rank
            ind[s[i].index] = i
        for i in range(n):
            nextindex = s[i].index + k // 2
            s[i].rank[1] = s[ind[nextindex]].rank[0] \
                if (nextindex < n) else -1
        s = sorted(s, key=lambda x: (x.rank[0], x.rank[1]))
        k *= 2
    res = [0] * n
    for i in range(n):
        res[i] = s[i].index
    return res


def buildLCP(t, suf):
    n = len(suf)
    inv = [0 for _ in range(n)]
    for i in range(n):
        inv[suf[i]] = i
    lcp = [0 for _ in range(n)]
    k = 0
    for i in range(n):
        if inv[i] == n - 1:
            k = 0
            continue
        j = suf[inv[i] + 1]
        while i + k < n and j + k < n and t[i + k] == t[j + k]:
            k += 1
        lcp[inv[i]] = k
        if k > 0:
            k -= 1
    return lcp


def main():
    s1 = input()
    s2 = input()
    t = s1 + "#" + s2 + "$"
    len1 = len(s1)
    suf = buildSuffix(t, len(t))
    print(suf)
    lcp = buildLCP(t, suf)
    print(lcp)
    q = []
    for i in range(2, len(lcp) - 1):
        if suf[i] > len1 > suf[i + 1] or suf[i] < len1 < suf[i + 1]:
            q.append((max(suf[i], suf[i + 1]),lcp[i]))
    q = sorted([(a-len1-1, b) for a, b in q if b])
    print(q)
    v = [0 for _ in range(len(s2) + 1)]
    res = len(s2)
    for a, b in q:
        if not v[b] and not v[b + a]:
            for i in range(a):
                v[b + i] = 1
            res -= (a - 1)
    return res


if __name__ == "__main__":
    print(main())

File: practice/ICPC2016Divisional/I.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <unordered_map>
using namespace std;

struct suffix{
    int index;
    pair<int,int> rank;
    suffix(int a, int b, int c){
        index = a;
        rank = {b,c};
    }
};

vector<int> buildSuffix(string txt, int n){
    vector<suffix> s;
    for (int i=0;i<n;i++){
        int rank0 = txt[i]-'A';
        int rank1 = (i+1<n) ? txt[i+1]-'A' : -1;
        suffix temp(i, rank0, rank1);
        s.push_back(temp);
    }
    sort(s.begin(),s.end(), [](const suffix &x, const suffix &y){ return (x.rank < y.rank);});
    vector<int> ind(n,0);
    int k = 4;
    while (k<2*n){
        int rank = 0;
        int prev_rank = s[0].rank.first;
        s[0].rank.first = rank;
        ind[s[0].index] = 0;
        for (int i=1;i<n;i++){
            if (s[i].rank.first == prev_rank && s[i].rank.second == s[i-1].rank.second){
                prev_rank = s[i].rank.first;
                s[i].rank.first = rank;
            }
            else{
                prev_rank = s[i].rank.first;
                rank += 1;
                s[i].rank.first = rank;
            }
            ind[s[i].index] = i;
        }
        int nextindex;
        for (int i=0;i<n;i++){
            nextindex = s[i].index + k/2;
            s[i].rank.second = (nextindex<n)?s[ind[nextindex]].rank.first:-1;
        }
        sort(s.begin(),s.end(), [](const suffix &x, const suffix &y){ return (x.rank < y.rank);});
        k *= 2;
    }
    vector<int> res(n,0);
    for (int i=0;i<n;i++){
        res[i] = s[i].index;
    }
    return res;
}

vector<int> buildLCP(string t, vector<int> suf){
    int n = suf.size();
    vector<int> inv(n,0);
    vector<int> lcp(n,0);
    for (int i=0;i<n;i++){
        inv[suf[i]] = i;
    }
    int k = 0;
    int j;
    for (int i=0;i<n;i++){
        if (inv[i] == n-1){
            k=0;
            continue;
        }
        j = suf[inv[i]+1];
        while (i+k<n && j+k<n && (t[i+k]==t[j+k])){
            k += 1;
        }
        lcp[inv[i]] = k;
        if (k>0){
            k -= 1;
        }
    }
    return lcp;
}

int dp(vector<int>& s, vector<int>& e, vector<int>& l) {
    map<int, int> times;
    unordered_map<int, vector<pair<int, int>>> j;
    for (auto i = 0; i < s.size(); ++i) {
        times[s[i]] = 0;
        j[s[i]].push_back({e[i],l[i]});
    }    
    int res = 0;
    for (auto it = rbegin(times); it != rend(times); ++it) {
        for (auto i : j[it->first]) {
            auto it = times.lower_bound(i.first);
            res = max(res, (it == end(times) ? 0 : it->second) + i.second);
        }
        it->second = res;
    }
    return res;
}  

int main(){
    string s1,s2;
    cin >> s1 >> s2;
    string t = s1 + '#' + s2 + '$';
    int len1 = s1.length();
    vector<int> suf = buildSuffix(t,t.length());
    vector<int> lcp = buildLCP(t, suf);
    vector<int> s;
    vector<int> e;
    vector<int> l;
    for (int i=0;i<suf.size();i++){
        cout << suf[i] << " ";
    }
    cout << "\n";
    for (int i=0;i<lcp.size();i++){
        cout << lcp[i] << " ";
    }
    
    for (int i=2;i<lcp.size()-1;i++){
        if ((suf[i]>len1 && suf[i+1] < len1) || (suf[i]<len1 && len1<suf[i+1])){
            // if (lcp[i]!=0){
                s.push_back(max(suf[i],suf[i+1])-len1-1);
                e.push_back(max(suf[i],suf[i+1])-len1-1+lcp[i]);
                l.push_back(lcp[i]-1);
            // }
        }
    }
    for (int i=0;i<s.size();i++){
        cout << "(" << s[i] << " " << e[i] << " " << l[i] << ") ";
    }
    vector<int> v(s2.length()+1,0);
    cout<<s2.length()-dp(s,e,l);
    return 0;
}